### ISSUE TITLE
utf8 match bug

### ISSUE THREAD

**Issue by sdeber** [01org/hyperscan#57]
I think there is a bug regarding to UTF8 patterns. Here are my tests with Chinese strings:

Test 1:
Pattern: 测试
Scan: 这是一个测试
Compilation flags: HS_FLAG_UTF8
Result: Success

Test 2: 
Pattern: \Q测试\E
Scan: 这是一个测试
Compilation flags: HS_FLAG_UTF8
Result: Failed

In Test 2, it should have passed, but instead it failed.

**Comment by coytea** [comment]
Hi sdeber,

I am trying to reproduce this issue on my machine but I am not seeing the failure for pattern 2. Are you able to attach your test program to the issue (if it is not too big)?

It looks to me if both versions of the pattern end up simply looking for the byte string \xe6\xb5\x8b\xe8\xaf\x95.

I will look further into our code and seeing if I can spot any suspicious interactions between our UTF8 handling and our parsing of \Q\E.

Also, are you using version 4.5 of Hyperscan or an older release?

Thanks,

Alex

**Comment by sdeber** [comment]
`#include <cstdlib>
#include <hs/hs.h>
#include <string>
#include <iostream>

using namespace std;

void test2()
{
    const char * pattern[] = {"\\Q测试\\E"};
    unsigned int ids[] = {1};
    unsigned int flags[] = {HS_FLAG_DOTALL | HS_FLAG_SINGLEMATCH | HS_FLAG_UTF8 | HS_FLAG_CASELESS};
    string content = u8"这是一个测试";

    cout << "content: " << content << endl;

    hs_database_t * db;
    hs_compile_error_t * perror;
    hs_error_t error = hs_compile_multi(pattern, flags, ids, 1, HS_MODE_BLOCK, NULL, &db, &perror);
    if (error != HS_SUCCESS)
    {
        cout << "Error code in compilation: " << error << ", " << perror->message << endl;
        return;
    }

    hs_scratch_t * pscratch = NULL;
    hs_alloc_scratch(db, &pscratch);
    
    if (pscratch == NULL)
        cout << "Error, null Scratch" << endl;

    error = hs_scan(db, content.c_str(), content.size(), 0, pscratch, &cb, NULL);
    cout << "Result code: " << error << endl;

}
int main(int argc, char** argv)
{

    test2();
    
    return 0;
}
`
Somehow, it failed to compile the pattern with error -4.

**Comment by sdeber** [comment]
[hyperscan.txt](https://github.com/01org/hyperscan/files/1079923/hyperscan.txt)

Somehow, it failed to compile the pattern with error code -4.

**Comment by coytea** [comment]
Thanks. It looks like HS_FLAG_CASELESS is required as well. I will track down what is going on with this.

**Comment by sdeber** [comment]
@coytea You are right. After I removed HS_FLAG_CASELESS, it actually parsed correctly and the matching also works

**Comment by coytea** [comment]
We tracked the issue to us incorrectly treating the contents of a \Q \E region as a sequence of bytes rather than a sequence of codepoints when in UTF8 mode. This resulted in the issue that you observed when the caseless flag was used.

We have pushed commits to the develop branch which fix the issue (a185be5a4f684c9bdbd90a2b9716ca02dde9e7b2 and d317d75615cdc6d0533e290de21a9bd205ffd12e). The fix will make it into the next release.

Thanks for spotting and reporting this bug.
